### ISSUE TITLE
prisma@4.7.1 - "RangeError: Maximum call stack size exceeded" when clientExtensions enabled

### ISSUE THREAD

**Issue by JoeKarow** [prisma/prisma#16600]
### Bug description

Enabling `clientExtensions` feature causes TypeScript failure.

```bash
RangeError: Maximum call stack size exceeded
    at getResolvedMembersOrExportsOfSymbol (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:48886:53)
    at getMembersOfSymbol (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:48932:19)
    at isEmptyAnonymousObjectType (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:54984:60)
    at Object.some (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:663:25)
    at getNormalizedUnionOrIntersectionType (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:55142:44)
    at getNormalizedType (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:55128:48)
    at isRelatedTo (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:55440:30)
    at typeRelatedToEachType (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:55751:35)
    at unionOrIntersectionRelatedTo (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:55687:28)
    at structuredTypeRelatedToWorker (/Users/joe/Library/pnpm/global/5/.pnpm/typescript@4.9.3/node_modules/typescript/lib/tsc.js:56024:34)
```

### How to reproduce

Please see notes on reproduction repository README: https://github.com/JoeKarow/prisma4.7.1-reproduce

### Prisma information

Prisma schema: https://github.com/JoeKarow/prisma4.7.1-reproduce/blob/main/prisma/schema.prisma

### Environment & setup

- OS: macOS 13.0.1 Intel
- Database: PostgreSQL
- Node.js version: 16.18.1

### Prisma Version

```bash
Environment variables loaded from .env
prisma                  : 4.7.1
@prisma/client          : 4.7.1
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 272861e07ab64f234d3ffc4094e32bd61775599c (at node_modules/.pnpm/@prisma+engines@4.7.1/node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli 272861e07ab64f234d3ffc4094e32bd61775599c (at node_modules/.pnpm/@prisma+engines@4.7.1/node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core 272861e07ab64f234d3ffc4094e32bd61775599c (at node_modules/.pnpm/@prisma+engines@4.7.1/node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt 272861e07ab64f234d3ffc4094e32bd61775599c (at node_modules/.pnpm/@prisma+engines@4.7.1/node_modules/@prisma/engines/prisma-fmt-darwin)
Format Wasm             : @prisma/prisma-fmt-wasm 4.7.1-1.272861e07ab64f234d3ffc4094e32bd61775599c
Default Engines Hash    : 272861e07ab64f234d3ffc4094e32bd61775599c
Studio                  : 0.477.0
Preview Features        : multiSchema, filteredRelationCount, fieldReference, postgresqlExtensions, clientExtensions
```

**Comment by tavindev** [comment]
Same here. Looks like a circular dependency thing.

**Comment by kachar** [comment]
Same here

```
Exception has occurred: RangeError: Maximum call stack size exceeded
  at getResolvedMembersOrExportsOfSymbol (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:59024:53)
    at getMembersOfSymbol (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:59079:19)
    at isEmptyAnonymousObjectType (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:65891:90)
    at Object.some (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:748:25)
    at getNormalizedUnionOrIntersectionType (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:66056:73)
    at getNormalizedType (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:66042:84)
    at isRelatedTo (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:66404:30)
    at isPropertySymbolTypeRelated (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:67746:24)
    at propertyRelatedTo (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:67789:31)
    at propertiesRelatedTo (/app/cells/graph-api/node_modules/typescript/lib/typescript.js:67990:43)
```

**Comment by JoeKarow** [comment]
@tavindev @kachar 
Are you able to link your schemas so the Prisma team has more data available to them  to be able to hunt down & fix this bug?

**Comment by kachar** [comment]
JoeKarow I believe @tavindev has shared his schema in the issue description - https://github.com/JoeKarow/prisma4.7.1-reproduce/blob/main/prisma/schema.prisma

You can find mine in: https://gist.github.com/kachar/77747ab867f99824e40480d041133029

**Comment by tavindev** [comment]
> @tavindev @kachar Are you able to link your schemas so the Prisma team has more data available to them to be able to hunt down & fix this bug?

Unfortunately, it will not be possible to share my schema.

**Comment by ajhollowayvrm** [comment]
I'm having this same error too. For me it only happens when I actually try to use an extension not just enabling it. 

`export const DBConn = new PrismaClient({
    log: [
        {
            emit: 'event',
            level: 'query',
        },
        {
            emit: 'stdout',
            level: 'error',
        },
        {
            emit: 'stdout',
            level: 'info',
        },
        {
            emit: 'stdout',
            level: 'warn',
        },
    ],
}).$extends({
    name: 'user display name',
    result: {
        users: {
            displayName: {
                needs: { first_name: true, preferred_name: true },
                compute(user) {
                    return user.preferred_name || user.first_name;
                },
            },
        },
    },
});`

If I remove the .$extension part, everything works just fine. Starting it with ts-node.

`[nodemon] watching extensions: ts,json
[nodemon] starting `ts-node ./src/api.ts`
/Users/ajholloway/programming/nectarbackend/node_modules/typescript/lib/typescript.js:66937
                    outofbandVarianceMarkerHandler = function (onlyUnreliable) {
                                                              ^
RangeError: Maximum call stack size exceeded
    at outofbandVarianceMarkerHandler (/Users/ajholloway/programming/nectarbackend/node_modules/typescript/lib/typescript.js:66937:63)`

**Comment by bn-piar** [comment]
Next.js app fails to run when using custom server with clientExtension feature. Without the custom server it runs fine locally but fails to build at digital ocean's app platform. 
Schema: https://pastebin.com/wPBPwvGE

**Comment by lipis** [comment]
I installed the 4.8.0 and still have the issue when building a Next.js app with `$extend`

**Comment by mattbajorek** [comment]
Same issue hear without even using `$extends` just upgrading to 4.8.0 crashes our compilation.

**Comment by millsp** [comment]
Hey @lipis @mattbajorek, sorry that you're still experiencing this. We've fixed and tested the original reproduction case, but it seems there is more to it. Could you please provide us with a reproduction?

**Comment by khalil-omer** [comment]
Confirmed this still exists in 4.8.0. `$extends` client method triggers this, as does `Prisma.defineExtension`.

**Comment by lipis** [comment]
@millsp in case you still need my use case... here is what I did with `$extend` in my Next.js project: https://gist.github.com/lipis/7564e99b6315a3940dbcb1a5c1a413a9

**Comment by millsp** [comment]
Hey @lipis @mattbajorek @khalil-omer, I have been working on this in a an experimental branch and as part of a wider effort. I would really like to get this right for you guys, could you please try `@prisma/client` `4.9.0-integration-feat-client-ext-generic-type-api.11`? On my side, I have tested it on a relatively huge & complex schema (~ +70 models).

**Comment by beeman** [comment]
I got a similar error when I tried to override an existing optional property:

I would love to use it like this, providing a default value based on another field

```ts
  readonly prisma = new PrismaClient().$extends({
    result: {
      user: {
        name: {
          needs: { name: true, username: true },
          compute: (parent) => parent.name ?? parent.username,
        },
      },
    },
  })
```

When I pick another name it works:

```ts
  readonly prisma = new PrismaClient().$extends({
    result: {
      user: {
        another_name: {
          needs: { name: true, username: true },
          compute: (parent) => parent.name ?? parent.username,
        },
      },
    },
  })
```

It would be nice if this use case would be supported, or give an error at compile time.


```
.../node_modules/@prisma/client/runtime/index.js:31227
const resolveNeeds = /* @__PURE__ */ __name((fieldName) => {
^
  RangeError: Maximum call stack size exceeded
  at resolveNeeds (.../node_modules/@prisma/client/runtime/index.js:31227:47)
  at Array.flatMap (<anonymous>)
  at .../node_modules/@prisma/client/runtime/index.js:31230:48
  at Cache.getOrCreate (.../node_modules/@prisma/client/runtime/index.js:31199:19)
  at resolveNeeds (.../node_modules/@prisma/client/runtime/index.js:31228:18)
  at Array.flatMap (<anonymous>)
  at .../node_modules/@prisma/client/runtime/index.js:31230:48
  at Cache.getOrCreate (.../node_modules/@prisma/client/runtime/index.js:31199:19)
  at resolveNeeds (.../node_modules/@prisma/client/runtime/index.js:31228:18)
  at Array.flatMap (<anonymous>)
```

**Comment by khalil-omer** [comment]
@millsp Hi - looks like you are making progress with this. With your branch installed, the call stack error is gone and `Prisma.defineExtension` throws no tsc errors. However, `$extends` is now throwing the following error. 

`Exported variable 'pivotPrisma' has or is using name 'Payload' from external module "[path]/client/runtime/index" but cannot be named.
`

The above error is actually throwing many times. Not sure if that fact helps you debug. 

![CleanShot 2022-12-27 at 09 47 36](https://user-images.githubusercontent.com/68179593/209703646-a64237de-caac-44a2-a306-05e961444e94.png)

**Comment by millsp** [comment]
Hey @beeman, I think we've had another report of [this here](https://github.com/prisma/prisma/issues/16643).

**Comment by millsp** [comment]
@khalil-omer Thanks, that's encouraging news. I had a quick try exporting a prisma client with an extension, but could not reproduce. Could you please share a minimal reproduction with me? If you have the time, I'd need a tsconfig and a small piece of code.

**Comment by khalil-omer** [comment]
@millsp this latest error appears to be related to `declarations: true` and `declerationMaps: true`. If I set both properties to `false` in the tsconfig, the error disappears and the build is successful. 

As someone commented [here](https://github.com/prisma/prisma/issues/16536#issuecomment-1333155610), there is an outstanding bug with declarations. Maybe there should be a separate issue created for that.

**Comment by Multiply** [comment]
> 4.9.0-integration-feat-client-ext-generic-type-api.11

It seems like nested includes doesn't generate correct typescript definitions with this branch.
```typescript
{
  include: {
    blocks: {
      include: {
        block: true,
      },
      orderBy: {
        sorting: 'asc',
      },
    },
    image: true,
  },
}
```

<img width="1206" alt="image" src="https://user-images.githubusercontent.com/453031/210184933-d5284c94-be48-47c7-a3a3-2cfcccce0484.png">

**Comment by millsp** [comment]
Thanks both, I was able to reproduce and thanks for the additional information.

**Comment by millsp** [comment]
Thanks to your help `4.9.0-integration-feat-client-ext-generic-type-api.13`, should fix the export issue as well as the node inference issue @khalil-omer. Could you please give it a try? I am not 100% for you @Multiply since I was not able to reproduce this exact same error. If the issue persists, would you mind sending me your schema?

Many thanks for the help so far :+1:

**Comment by jove4015** [comment]
I just ran into this problem today trying to move RLS from middleware into client extensions.  The code actually works fine in development mode - but I can't make a production build because of the above error.

I tried using your branch 4.9.0-integration-feat-client-ext-generic-type-api.13, and while the recursion error goes away, I have errors all over my project now:
![image](https://user-images.githubusercontent.com/228216/212200048-53a9f483-ab94-4283-9534-3ec2a714cbaa.png)

All in the same vein where the types of everything are just "never".

On a hunch I tried v.14, 15, 16, and 17 (which seems to be the latest one available) as well since I figured you were still working on this, but I'm still getting the same errors with "never".

Still, in dev mode, my app loads and functions properly.  It's just a type resolution issue.

**Comment by TaylorFacen** [comment]
Hi, I'm running into this error as well. Using the 4.9.0 branch resolved the maximum call stack error, but now my nested values aren't typed. For example

```
Type error: Property 'post' does not exist on type '{ id: string; createdAt: Date; updatedAt: Date; postId: string; commentBody: string; }'.
```

Even though there aren't any type errors visible in VScode's linting and I am in fact querying the comment and the post
e.g.
```
prisma.comment.findFirstOrThrow({ where: { id: commentId }, include: { post: true } })
```

**Comment by millsp** [comment]
Hey @TaylorFacen, thanks for the reporr. Could you please create a separate issue?

**Comment by TaylorFacen** [comment]
> Hey @TaylorFacen, thanks for the report. Could you please create a separate issue?

Done! https://github.com/prisma/prisma/issues/17349 . Let me know if you need any additional information.

**Comment by jove4015** [comment]
@millsp  - This bug is marked resolved, however, client extensions still are not usable (see https://github.com/prisma/prisma/issues/17767).  Is there any progress on making this feature functionally available?

**Comment by WARDHAN10** [comment]
Can't seems to follow what is the actuals solution to this can anybody elaborate, getting the same issue

**Comment by millsp** [comment]
Hey @WARDHAN10, we have fixed this issue already, and the fix is available. Please update to our latest version, and if the problem persists, we'd love if you could open a new issue with a reproduction. Thank you :)
